bigquery_etl/monitoring/export_metrics.py

```diff
@@ -0,0 +1,216 @@
+    alignment_period: int = 300,
+):
+    """
+    Fetch given metric from cloud monitoring and write results to designated table.
+
+    :param monitoring_project: Project containing the monitoring workspace
+    :param dst_project: Project to write results to
+    :param dst_dataset: Bigquery dataset to write results to
+    :param dst_table: Bigquery table to write results to
+    :param metric: Metric and resource identifier
+        (e.g. pubsub.googleapis.com/topic/send_request_count)
+    :param execution_time: End of the time interval to export
+    :param time_offset: Number of hours to offset interval to account for delayed metrics
+    :param interval_hours: Number of hours to export, ending at execution time
+    :param overwrite: Overwrite destination table (default is append)
+    :param aggregator: Cloud monitoring series aggregator
+    :param aligner: Cloud monitoring series aligner
+    :param alignment_period: Seconds between each point in the time series
+    """
+    monitoring_client = monitoring.MetricServiceClient()
+
+    bq_client = bigquery.Client(dst_project)
+
+    # Force timezone to be UTC so intervals match exported metrics
+    os.environ["TZ"] = "UTC"
+    time.tzset()
+
+    # Larger intervals cause run time to increase superlinearly and takes a lot of memory
+    max_interval_width = 24
+    for interval_offset in range(0, interval_hours, max_interval_width):
+        offset = time_offset + interval_offset
+        end_time = execution_time - datetime.timedelta(hours=offset)
+        start_time = execution_time - datetime.timedelta(
+            hours=min(offset + max_interval_width, time_offset + interval_hours)
+        )
+        print(f"Fetching values for {start_time} to {end_time}")
+
+        metric_filter = f'metric.type = "{metric}"'
+
+        aggregation = monitoring.Aggregation(
+            {
+                "alignment_period": {"seconds": alignment_period},
+                "cross_series_reducer": aggregator,
+                "per_series_aligner": aligner,
+            }
+        )
+
+        time_series_data = get_time_series(
+            monitoring_client,
+            monitoring_project,
+            metric_filter,
+            aggregation,
+            start_time,
+            end_time,
+        )
+
+        print(
+            f"Retrieved {len(time_series_data)} data points for"
+            f" {start_time} to {end_time}"
+        )
+
+        target_dataset = bigquery.DatasetReference(dst_project, dst_dataset)
+        target_table = bigquery.table.TableReference(target_dataset, dst_table)
+
+        # get existing timestamps in destination table to avoid overlap
+        if not overwrite and len(time_series_data) > 0:
+            time_series_data = filter_existing_data(
+                time_series_data,
+                bq_client,
+                target_table,
+                start_time,
+                end_time,
+            )
```

Comment:
My inclination would be to create these monitoring tables with hourly partitioning, and then have these scripts atomically overwrite the target partition (specifying the destination table as `mytable$2021011201`, etc.), avoiding the need for filtering logic like this.

We could achieve some simplification by having all this machinery assume that it's operating on one whole hour at a time. I may well be missing some nuance, though, so definitely open to pushback.